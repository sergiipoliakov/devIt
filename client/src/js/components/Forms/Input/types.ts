import { ReactNode } from 'react';

export interface InputProps {
    id?: string;
    dataCy?: string;
    name?: string;
    disabled?: boolean;
    value?: string | number | null;
    label?: string;
    type?: string;
    sufix?: string | ReactNode;
    autocomplete?: string;
    allowClear?: boolean;
    icon?: {
        icon?: string,
        size?: string,
        color?: string,
        dataCy?: string,
        classNames?: {
            wrapper?: string,
            icon?: string
        }
        onClickIconHandler?: () => void
    };
    meta?: {
        error?: string,
        touched?: boolean,
        warning?: string,
    };
    placeholder?: string;
    classNames?: {
        wrapper?: string;
        label?: string;
        field?: string;
    };
    prefix?: ReactNode
    onChange?: (field: string, value: any) => void,
    onBlur?: (value: string) => void,
}
