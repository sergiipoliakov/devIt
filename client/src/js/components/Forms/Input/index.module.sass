@import '../default.vars.sass'
@import '../../../../styles/__modificators/__text.sass'

.input
    position: relative
    border: 1px solid $form-element-default-color
    padding: $form-element-padding
    color: $form-element-font-color
    border-width: $form-element-border-width
    border-color: $form-element-border-color
    border-radius: $form-element-border-radius
    transition: $form-element-transition
    background-color: $form-element-background
    @extend %text--default

    &__icon
        position: absolute
        z-index: 2
        right: 16px
        bottom: 50%
        transform: translateY(50%)
    &:focus,
    &:hover
        border-color: $form-element-active-border-color
        box-shadow: none
        border-inline-end-width: $form-element-border-inline-end-width
    &--error
        border-color: $form-element-error-color !important
    &--sufix
        padding-right: 48px
    & input 
        &:disabled
            cursor: not-allowed
            color: $form-element-disabled-font-color
            background-color: $form-element-disabled-background
            border-color: $form-element-default-color !important
            -webkit-text-fill-color: $form-element-disabled-font-color
            opacity: 1
            border-inline-end-width: $form-element-border-inline-end-width !important
    &:disabled
        cursor: not-allowed
        color: $form-element-disabled-font-color
        background-color: $form-element-disabled-background
        border-color: $form-element-default-color !important
        -webkit-text-fill-color: $form-element-disabled-font-color
        opacity: 1
        border-inline-end-width: $form-element-border-inline-end-width !important
    &::placeholder
        color: $form-element-placeholder-color
        font-weight: $form-element-placeholder-font-weight
    & input::placeholder
        color: $form-element-placeholder-color

    &__message
        width: 100%
        position: absolute
        left: 4px
        bottom: -22px
        &--error
            color: $form-element-error-color 

    &__sufix
        position: absolute
        right: 16px
        bottom: 50%
        transform: translateY(50%)
        z-index: 2
