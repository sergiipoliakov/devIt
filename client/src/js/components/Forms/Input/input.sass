@import '../default.vars.sass'

#wi
    .ant-input-outlined.ant-input-disabled
        background-color: $form-element-disabled-background
