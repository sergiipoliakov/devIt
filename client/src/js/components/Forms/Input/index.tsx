import React from 'react';
import Input from 'antd/lib/input';
import Label from '../Label';
// import Icons from '../../Icons';
import Text from '../../Text';
import { InputProps } from './types';
import inputStyles from './index.module.sass';
import './input.sass';

const InputComponent = (props: InputProps): JSX.Element => {
    const {
        id,
        dataCy,
        value,
        name,
        label,
        meta,
        disabled,
        sufix,
        autocomplete = 'off',
        type,
        prefix,
        allowClear,
        // icon: {
        //     icon,
        //     size: iconSize = '16',
        //     classNames: {
        //         wrapper: iconWrapperClassName = '',
        //         icon: iconClassName = ''
        //     } = {},
        //     color: iconColor = 'default',
        //     onClickIconHandler,
        //     dataCy: iconDataCy
        // } = {},
        placeholder,
        classNames: {
            label: labelClass = '',
            wrapper: wrapperClass = '',
            field: fieldClass = ''
        } = {},
        onChange
    } = props;
    const htmlFor = `input--f${(~~(Math.random() * 1e8)).toString(16)}`;
    const { touched, error, warning } = meta || {};

    const classList = [fieldClass || '', inputStyles.input || '', ...(touched && (error || warning) ? [inputStyles['input--error']] : []), ...(sufix ? [inputStyles['input--sufix']] : [])].join(' ');

    const onChangeValue = (event: React.ChangeEvent<HTMLInputElement>) => {
        onChange?.(name as string, event?.currentTarget?.value);
    };

    const setError = () => {
        if (touched && (error || warning)) {
            return (
                <>
                    {touched && (error || warning) ? (
                        <div className={`${inputStyles.input__message} ${inputStyles['input__message--error']} fl overflow--hidden`}>
                            {/* <Icons
                                dataCy="error-icon"
                                icon="flat/warning"
                                size="16"
                                color="error"
                                classNames={{
                                    wrapper: 'margin--r-8'
                                }}
                            /> */}
                            <Text
                                size="small"
                                dataCy="input-error-text"
                                color="error"
                                weight="400"
                                className="text-overflow--ellipsis overflow--hidden font--no-wrap"
                            >
                                {(error || warning)}
                            </Text>
                        </div>
                    ) : null}
                </>
            );
        }
        return null;
    };

    return (
        <div className={`fl ${wrapperClass} fl--dir-col width--100 relative--core`}>
            {label ? (
                <Label
                    htmlFor={id || htmlFor}
                    label={label}
                    className={`${labelClass || ''}`}
                />
            ) : null}
            <div className="relative--core width--100">
                {/* {icon ? (
                    <Icons
                        dataCy={iconDataCy}
                        icon={icon}
                        size={iconSize}
                        color={iconColor}
                        classNames={{
                            wrapper: `${inputStyles.input__icon} ${iconWrapperClassName || ''}`,
                            ...(iconClassName ? { icon: iconClassName } : {})
                        }}
                        {...(onClickIconHandler ? { onClick: onClickIconHandler } : {})}
                    />
                ) : null} */}
                {sufix ? (
                    <Text className={`${inputStyles.input__sufix}`} color="default" weight="400" transform="uppercase">{sufix}</Text>
                ) : null}

                <Input
                    data-cy={`${dataCy}-input`}
                    id={id || htmlFor}
                    type={type}
                    name={name}
                    value={value as any}
                    autoComplete={autocomplete}
                    className={`${classList} width--100`}
                    onChange={onChangeValue}
                    disabled={disabled}
                    placeholder={placeholder}
                    allowClear={allowClear}
                    {...(prefix ? { prefix } : {})}
                />
            </div>
            {setError()}
        </div>
    );
};

export default InputComponent;
