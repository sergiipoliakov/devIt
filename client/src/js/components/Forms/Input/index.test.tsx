import React from 'react';
import {
 render, waitFor
} from '@testing-library/react';
import { fireEvent } from '@testing-library/dom';

import Input from '.';

jest.mock('react', () => ({
    ...jest.requireActual('react')
}));

const setup = () => {
    const wrapper = render(
        <Input 
            id="test-input" 
            type="text" 
            label="Test label" 
            classNames={{
                label: 'test-label-class',
                wrapper: 'wrapper-input-class',
                field: 'field-input-class'
            }}
            onChange={() => 'UnitTest'} 
        />
    );
    const { container } = wrapper;
    const input = container.querySelector('#test-input');

    return {
        wrapper, 
        input
    };
};

describe('Input Component', () => {
    it('Snapshot testing', () => {
        const { wrapper } = setup();
        expect(wrapper.asFragment()).toMatchSnapshot();
    });
    it('Changing value testing', async () => {
        const mock = {
            testValue: '9FyD3Mqs9H3Md0P!@0@4'
        };
        const { input } = setup();
        fireEvent.change(input as any, { target: { value: mock.testValue } });

        await waitFor(() => {
            expect((input as any).value).toBe(mock.testValue);
        });
    });
});
