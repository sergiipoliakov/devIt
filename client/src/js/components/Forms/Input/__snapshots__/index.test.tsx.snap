// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="fl wrapper-input-class fl--dir-col width--100 relative--core"
  >
    <label
      class="label test-label-class"
      for="test-input"
    >
      Test label
    </label>
    <div
      class="relative--core width--100"
    >
      <input
        autocomplete="off"
        class="ant-input field-input-class input width--100"
        id="test-input"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
